pubnub/event_engine/effects.py

```diff
@@ -0,0 +1,431 @@
+        if hasattr(self, 'task') and isinstance(self.task, asyncio.Task) and not self.task.cancelled():
+            self.task.cancel()
+
+    def get_new_stop_event(self):
+        event = asyncio.Event()
+        self.logger.debug(f'creating new stop_event({id(event)}) for {self.__class__.__name__}')
+        return event
+
+    def calculate_reconnection_delay(self, attempts):
+        if self.reconnection_policy is PNReconnectionPolicy.LINEAR:
+            delay = self.interval
+
+        elif self.reconnection_policy is PNReconnectionPolicy.EXPONENTIAL:
+            delay = int(math.pow(2, attempts - 5 * math.floor((attempts - 1) / 5)) - 1)
+        return delay
+
+
+class HandshakeEffect(Effect):
+    def run(self):
+        channels = self.invocation.channels
+        groups = self.invocation.groups
+        tt = self.invocation.timetoken or 0
+        if hasattr(self.pubnub, 'event_loop'):
+            self.stop_event = self.get_new_stop_event()
+            self.run_async(self.handshake_async(channels=channels,
+                                                groups=groups,
+                                                timetoken=tt,
+                                                stop_event=self.stop_event))
+
+    async def handshake_async(self, channels, groups, stop_event, timetoken: int = 0):
+        request = Subscribe(self.pubnub).channels(channels).channel_groups(groups).cancellation_event(stop_event)
+
+        if feature_enabled('PN_MAINTAIN_PRESENCE_STATE') and hasattr(self.pubnub, 'state_container'):
+            state_container = self.pubnub.state_container
+            request.state(state_container.get_state(channels))
+
+        request.timetoken(0)
+        response = await request.future()
+
+        if isinstance(response, PubNubException):
+            self.logger.warning(f'Handshake failed: {str(response)}')
+            handshake_failure = events.HandshakeFailureEvent(str(response), 1, timetoken=timetoken)
+            self.event_engine.trigger(handshake_failure)
+        elif response.status.error:
+            self.logger.warning(f'Handshake failed: {response.status.error_data.__dict__}')
+            handshake_failure = events.HandshakeFailureEvent(response.status.error_data, 1, timetoken=timetoken)
+            self.event_engine.trigger(handshake_failure)
+        else:
+            cursor = response.result['t']
+            timetoken = timetoken if timetoken > 0 else cursor['t']
+            region = cursor['r']
+            handshake_success = events.HandshakeSuccessEvent(timetoken, region)
+            self.event_engine.trigger(handshake_success)
+
+
+class ReceiveMessagesEffect(Effect):
+    invocation: invocations.ReceiveMessagesInvocation
+
+    def run(self):
+        channels = self.invocation.channels
+        groups = self.invocation.groups
+        timetoken = self.invocation.timetoken
+        region = self.invocation.region
+
+        if hasattr(self.pubnub, 'event_loop'):
```

Comment:
I'm curious, what if there's no `event_loop` attribute in `self.pubnub`? 
What will happen then?

Reply:
It's a switch for case when you want to use it with synchronous version of PN. I want to leave it in case there will be thread based implementation in the future